pkg/app/piped/planpreview/builder.go

```diff
@@ -158,82 +160,121 @@ func (b *builder) build(ctx context.Context, id string, cmd model.Command_BuildP
 	}
 	results := failedResults
 
+	if len(triggerApps) == 0 {
+		return results, nil
+	}
+
 	// Plan the trigger applications for more detailed feedback.
-	for _, app := range triggerApps {
-		b.logger.Info("will decide sync strategy for an application",
-			zap.String("id", app.Id),
-			zap.String("name", app.Name),
-			zap.String("kind", app.Kind.String()),
-		)
+	var (
+		numApps  = len(triggerApps)
+		appCh    = make(chan *model.Application, numApps)
+		resultCh = make(chan *model.ApplicationPlanPreviewResult, numApps)
+	)
+	numWorkers := defaultAppWorkerNum
+	if numWorkers > numApps {
+		numWorkers = numApps
+	}
 
-		// We only need the environment name
-		// so the returned error can be ignorable.
-		var envName string
-		if env, err := b.environmentGetter.Get(ctx, app.EnvId); err == nil {
-			envName = env.Name
-		}
+	// Start some workers to speed up building time.
+	b.logger.Info(fmt.Sprintf("start %d workers for building plan-preview results for %d applications", numWorkers, numApps))
```

Comment:
should we use the above wrapped `logger`? 👀

Reply:
Right. Nice catch!